File: crypto_screener/derivatives.py
from __future__ import annotations

from typing import Any

from .scoring import clamp, mean, pct_change, stdev, to_float


INTERVAL_HOURS = {
    "1m": 1.0 / 60.0,
    "3m": 3.0 / 60.0,
    "5m": 5.0 / 60.0,
    "15m": 15.0 / 60.0,
    "30m": 0.5,
    "1h": 1.0,
    "4h": 4.0,
    "6h": 6.0,
    "8h": 8.0,
    "12h": 12.0,
    "1d": 24.0,
    "1w": 24.0 * 7.0,
}


def interval_hours(interval: str) -> float:
    return INTERVAL_HOURS.get(interval, 24.0)


def candles_per_window(interval: str, hours: float) -> int:
    return max(1, round(hours / interval_hours(interval)))


def derivatives_snapshot(
    oi_history: list[dict[str, Any]],
    funding_history: list[dict[str, Any]],
    liquidation_history: list[dict[str, Any]],
    taker_history: list[dict[str, Any]],
    interval: str,
    end_time: int | None = None,
) -> dict[str, Any]:
    oi_rows = _series_until(_normalize_close_series(oi_history), end_time)
    funding_rows = _series_until(_normalize_close_series(funding_history), end_time)
    liquidation_rows = _series_until(_normalize_liquidations(liquidation_history), end_time)
    taker_rows = _series_until(_normalize_taker(taker_history), end_time)

    if not any((oi_rows, funding_rows, liquidation_rows, taker_rows)):
        return {}

    window = candles_per_window(interval, 24.0)
    oi_closes = [row["close"] for row in oi_rows]
    funding_closes = [row["close"] for row in funding_rows]
    liq_window = liquidation_rows[-window:]
    taker_window = taker_rows[-window:]

    oi_change_1 = _pct_change_steps(oi_closes, 1)
    oi_change_window = _pct_change_steps(oi_closes, window)
    oi_previous_change = _pct_change_steps(oi_closes[:-1], 1)
    oi_acceleration = (
        oi_change_1 - oi_previous_change
        if oi_change_1 is not None and oi_previous_change is not None
        else None
    )
    oi_zscore = _latest_zscore(oi_closes, 30)

    funding_avg = mean(funding_closes[-window:]) if funding_closes else None
    funding_abs_avg = mean([abs(value) for value in funding_closes[-window:]]) if funding_closes else None
    funding_persistence = _sign_persistence(funding_closes[-window:])

    long_liq = sum(row["long"] for row in liq_window)
    short_liq = sum(row["short"] for row in liq_window)
    liq_total = long_liq + short_liq
    liq_imbalance = ((short_liq - long_liq) / liq_total * 100.0) if liq_total > 0 else None

    buy_volume = sum(row["buy"] for row in taker_window)
    sell_volume = sum(row["sell"] for row in taker_window)
    taker_total = buy_volume + sell_volume
    taker_ratio = (buy_volume / sell_volume) if sell_volume > 0 else None
    taker_imbalance = ((buy_volume - sell_volume) / taker_total * 100.0) if taker_total > 0 else None

    confirmation = _derivatives_confirmation(oi_acceleration, taker_imbalance, liq_imbalance)

    result = {
        "derivatives_interval": interval,
        "derivatives_oi_count": len(oi_rows),
        "derivatives_funding_count": len(funding_rows),
        "derivatives_liquidation_count": len(liquidation_rows),
        "derivatives_taker_count": len(taker_rows),
        "oi_change_4h_pct_history": oi_change_1,
        "oi_change_24h_pct_history": oi_change_window,
        "oi_acceleration_4h_pct": oi_acceleration,
        "oi_zscore_30": oi_zscore,
        "funding_avg_24h_pct": funding_avg,
        "funding_abs_avg_24h_pct": funding_abs_avg,
        "funding_persistence_24h": funding_persistence,
        "long_liquidation_usd_24h_history": long_liq if liq_window else None,
        "short_liquidation_usd_24h_history": short_liq if liq_window else None,
        "liquidation_total_24h_usd": liq_total if liq_window else None,
        "liquidation_imbalance_24h_pct": liq_imbalance,
        "taker_buy_volume_usd_24h": buy_volume if taker_window else None,
        "taker_sell_volume_usd_24h": sell_volume if taker_window else None,
        "taker_buy_sell_ratio_24h": taker_ratio,
        "taker_imbalance_24h_pct": taker_imbalance,
        "derivatives_confirmation_score": confirmation,
    }
    return {key: value for key, value in result.items() if value is not None}


def _normalize_close_series(rows: list[dict[str, Any]]) -> list[dict[str, float]]:
    normalized: list[dict[str, float]] = []
    for row in sorted(rows, key=lambda item: to_float(item.get("time"), 0.0) or 0.0):
        time = to_float(row.get("time"))
        close = to_float(row.get("close"))
        if time is None or close is None:
            continue
        normalized.append({"time": time, "close": close})
    return normalized


def _normalize_liquidations(rows: list[dict[str, Any]]) -> list[dict[str, float]]:
    normalized: list[dict[str, float]] = []
    for row in sorted(rows, key=lambda item: to_float(item.get("time"), 0.0) or 0.0):
        time = to_float(row.get("time"))
        long_value = to_float(row.get("aggregated_long_liquidation_usd"), 0.0) or 0.0
        short_value = to_float(row.get("aggregated_short_liquidation_usd"), 0.0) or 0.0
        if time is None:
            continue
        normalized.append({"time": time, "long": long_value, "short": short_value})
    return normalized


def _normalize_taker(rows: list[dict[str, Any]]) -> list[dict[str, float]]:
    normalized: list[dict[str, float]] = []
    for row in sorted(rows, key=lambda item: to_float(item.get("time"), 0.0) or 0.0):
        time = to_float(row.get("time"))
        buy = to_float(row.get("aggregated_buy_volume_usd"), 0.0) or 0.0
        sell = to_float(row.get("aggregated_sell_volume_usd"), 0.0) or 0.0
        if time is None:
            continue
        normalized.append({"time": time, "buy": buy, "sell": sell})
    return normalized


def _series_until(rows: list[dict[str, float]], end_time: int | None) -> list[dict[str, float]]:
    if end_time is None:
        return rows
    return [row for row in rows if row["time"] <= end_time]


def _pct_change_steps(values: list[float], steps: int) -> float | None:
    if len(values) <= steps:
        return None
    return pct_change(values[-steps - 1], values[-1])


def _latest_zscore(values: list[float], window: int) -> float | None:
    if len(values) < max(3, window):
        return None
    sample = values[-window:]
    deviation = stdev(sample)
    if deviation == 0:
        return 0.0
    return (sample[-1] - mean(sample)) / deviation


def _sign_persistence(values: list[float]) -> float | None:
    if not values:
        return None
    signs = [1.0 if value > 0 else -1.0 if value < 0 else 0.0 for value in values]
    return mean(signs)


def _derivatives_confirmation(
    oi_acceleration: float | None,
    taker_imbalance: float | None,
    liquidation_imbalance: float | None,
) -> float | None:
    components: list[float] = []
    if oi_acceleration is not None:
        components.append(clamp(oi_acceleration / 8.0, -1.0, 1.0))
    if taker_imbalance is not None:
        components.append(clamp(taker_imbalance / 20.0, -1.0, 1.0))
    if liquidation_imbalance is not None:
        components.append(clamp(liquidation_imbalance / 60.0, -1.0, 1.0))
    return mean(components) if components else None

File: apps/api/src/pipeline/derivatives.ts
import { clamp, mean, pctChange, stdev, toFloat } from './scoring.js';

export type RawHistoryRow = Record<string, unknown>;

const INTERVAL_HOURS: Record<string, number> = {
  '1m': 1.0 / 60.0,
  '3m': 3.0 / 60.0,
  '5m': 5.0 / 60.0,
  '15m': 15.0 / 60.0,
  '30m': 0.5,
  '1h': 1.0,
  '4h': 4.0,
  '6h': 6.0,
  '8h': 8.0,
  '12h': 12.0,
  '1d': 24.0,
  '1w': 24.0 * 7.0,
};

function intervalHours(interval: string): number {
  return INTERVAL_HOURS[interval] ?? 24.0;
}

export function candlesPerWindow(interval: string, hours: number): number {
  return Math.max(1, Math.round(hours / intervalHours(interval)));
}

interface CloseRow {
  time: number;
  close: number;
}

interface LiquidationRow {
  time: number;
  long: number;
  short: number;
}

interface TakerRow {
  time: number;
  buy: number;
  sell: number;
}

// Keys whose value is null are dropped from the result entirely, not kept as null.
export function derivativesSnapshot(
  oiHistory: RawHistoryRow[],
  fundingHistory: RawHistoryRow[],
  liquidationHistory: RawHistoryRow[],
  takerHistory: RawHistoryRow[],
  interval: string,
  endTime?: number,
): Record<string, unknown> {
  const oiRows = seriesUntil(normalizeCloseSeries(oiHistory), endTime);
  const fundingRows = seriesUntil(normalizeCloseSeries(fundingHistory), endTime);
  const liquidationRows = seriesUntil(normalizeLiquidations(liquidationHistory), endTime);
  const takerRows = seriesUntil(normalizeTaker(takerHistory), endTime);

  if (
    oiRows.length === 0 &&
    fundingRows.length === 0 &&
    liquidationRows.length === 0 &&
    takerRows.length === 0
  ) {
    return {};
  }

  const window = candlesPerWindow(interval, 24.0);
  const oiCloses = oiRows.map((row) => row.close);
  const fundingCloses = fundingRows.map((row) => row.close);
  const liqWindow = liquidationRows.slice(-window);
  const takerWindow = takerRows.slice(-window);

  const oiChange1 = pctChangeSteps(oiCloses, 1);
  const oiChangeWindow = pctChangeSteps(oiCloses, window);
  const oiPreviousChange = pctChangeSteps(oiCloses.slice(0, -1), 1);
  const oiAcceleration =
    oiChange1 !== null && oiPreviousChange !== null ? oiChange1 - oiPreviousChange : null;
  const oiZscore = latestZscore(oiCloses, 30);

  const fundingWindow = fundingCloses.slice(-window);
  const fundingAvg = fundingCloses.length ? mean(fundingWindow) : null;
  const fundingPersistence = signPersistence(fundingWindow);

  const longLiq = liqWindow.reduce((sum, row) => sum + row.long, 0);
  const shortLiq = liqWindow.reduce((sum, row) => sum + row.short, 0);
  const liqTotal = longLiq + shortLiq;
  const liqImbalance = liqTotal > 0 ? ((shortLiq - longLiq) / liqTotal) * 100.0 : null;

  const buyVolume = takerWindow.reduce((sum, row) => sum + row.buy, 0);
  const sellVolume = takerWindow.reduce((sum, row) => sum + row.sell, 0);
  const takerTotal = buyVolume + sellVolume;
  const takerRatio = sellVolume > 0 ? buyVolume / sellVolume : null;
  const takerImbalance = takerTotal > 0 ? ((buyVolume - sellVolume) / takerTotal) * 100.0 : null;

  const confirmation = derivativesConfirmation(oiAcceleration, takerImbalance, liqImbalance);

  const result: Record<string, unknown> = {
    derivatives_interval: interval,
    derivatives_oi_count: oiRows.length,
    derivatives_funding_count: fundingRows.length,
    derivatives_liquidation_count: liquidationRows.length,
    derivatives_taker_count: takerRows.length,
    oi_change_4h_pct_history: oiChange1,
    oi_change_24h_pct_history: oiChangeWindow,
    oi_acceleration_4h_pct: oiAcceleration,
    oi_zscore_30: oiZscore,
    funding_avg_24h_pct: fundingAvg,
    funding_persistence_24h: fundingPersistence,
    long_liquidation_usd_24h_history: liqWindow.length ? longLiq : null,
    short_liquidation_usd_24h_history: liqWindow.length ? shortLiq : null,
    liquidation_imbalance_24h_pct: liqImbalance,
    taker_buy_sell_ratio_24h: takerRatio,
    taker_imbalance_24h_pct: takerImbalance,
    derivatives_confirmation_score: confirmation,
  };

  return Object.fromEntries(Object.entries(result).filter(([, value]) => value !== null));
}

export function sortByTime(rows: RawHistoryRow[]): RawHistoryRow[] {
  return [...rows].sort((a, b) => (toFloat(a.time, 0.0) ?? 0.0) - (toFloat(b.time, 0.0) ?? 0.0));
}

function normalizeCloseSeries(rows: RawHistoryRow[]): CloseRow[] {
  const normalized: CloseRow[] = [];
  for (const row of sortByTime(rows)) {
    const time = toFloat(row.time);
    const close = toFloat(row.close);
    if (time === null || close === null) {
      continue;
    }
    normalized.push({ time, close });
  }
  return normalized;
}

function normalizeLiquidations(rows: RawHistoryRow[]): LiquidationRow[] {
  const normalized: LiquidationRow[] = [];
  for (const row of sortByTime(rows)) {
    const time = toFloat(row.time);
    const long = toFloat(row.aggregated_long_liquidation_usd, 0.0) ?? 0.0;
    const short = toFloat(row.aggregated_short_liquidation_usd, 0.0) ?? 0.0;
    if (time === null) {
      continue;
    }
    normalized.push({ time, long, short });
  }
  return normalized;
}

function normalizeTaker(rows: RawHistoryRow[]): TakerRow[] {
  const normalized: TakerRow[] = [];
  for (const row of sortByTime(rows)) {
    const time = toFloat(row.time);
    const buy = toFloat(row.aggregated_buy_volume_usd, 0.0) ?? 0.0;
    const sell = toFloat(row.aggregated_sell_volume_usd, 0.0) ?? 0.0;
    if (time === null) {
      continue;
    }
    normalized.push({ time, buy, sell });
  }
  return normalized;
}

function seriesUntil<T extends { time: number }>(rows: T[], endTime: number | undefined): T[] {
  if (endTime === undefined) {
    return rows;
  }
  return rows.filter((row) => row.time <= endTime);
}

function pctChangeSteps(values: number[], steps: number): number | null {
  if (values.length <= steps) {
    return null;
  }
  return pctChange(values[values.length - steps - 1] as number, values.at(-1) as number);
}

function latestZscore(values: number[], window: number): number | null {
  if (values.length < Math.max(3, window)) {
    return null;
  }
  const sample = values.slice(-window);
  const deviation = stdev(sample);
  if (deviation === 0) {
    return 0.0;
  }
  return ((sample.at(-1) as number) - mean(sample)) / deviation;
}

function signPersistence(values: number[]): number | null {
  if (values.length === 0) {
    return null;
  }
  const signs = values.map((value) => (value > 0 ? 1.0 : value < 0 ? -1.0 : 0.0));
  return mean(signs);
}

function derivativesConfirmation(
  oiAcceleration: number | null,
  takerImbalance: number | null,
  liquidationImbalance: number | null,
): number | null {
  const components: number[] = [];
  if (oiAcceleration !== null) {
    components.push(clamp(oiAcceleration / 8.0, -1.0, 1.0));
  }
  if (takerImbalance !== null) {
    components.push(clamp(takerImbalance / 20.0, -1.0, 1.0));
  }
  if (liquidationImbalance !== null) {
    components.push(clamp(liquidationImbalance / 60.0, -1.0, 1.0));
  }
  return components.length ? mean(components) : null;
}
